src/jmh/java/com/lmax/disruptor/SimpleBenchmark.java

```diff
@@ -0,0 +1,80 @@
+package com.lmax.disruptor;
+
+import com.lmax.disruptor.dsl.Disruptor;
+import com.lmax.disruptor.dsl.ProducerType;
+import com.lmax.disruptor.util.Constants;
+import com.lmax.disruptor.util.DaemonThreadFactory;
+import com.lmax.disruptor.util.SimpleEvent;
+import com.lmax.disruptor.util.SimpleEventHandler;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OperationsPerInvocation;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+import java.util.concurrent.TimeUnit;
+
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@State(Scope.Thread)
+@Fork(1)
+public class SimpleBenchmark
+{
+    private RingBuffer<SimpleEvent> ringBuffer;
+    private SimpleEventHandler eventHandler;
+    private Disruptor<SimpleEvent> disruptor;
+
+    @Setup
+    public void setup() {
+        disruptor = new Disruptor<>(
+                SimpleEvent::new,
+                Constants.RINGBUFFER_SIZE,
+                DaemonThreadFactory.INSTANCE,
+                ProducerType.SINGLE,
+                new BusySpinWaitStrategy());
+
+        eventHandler = new SimpleEventHandler();
+        disruptor.handleEventsWith(eventHandler);
+
+        ringBuffer = disruptor.start();
+    }
+
+    @Benchmark
+    @OperationsPerInvocation(Constants.ITERATIONS)
+    public void publishSimpleEvents() {
+        for (int i = 0; i < Constants.ITERATIONS; i++) {
+            long sequence = ringBuffer.next();
+            SimpleEvent simpleEvent = ringBuffer.get(sequence);
+            simpleEvent.setValue(i);
+            ringBuffer.publish(sequence);
+        }
+
+        while (eventHandler.lastSeenSequence % Constants.ITERATIONS != Constants.ITERATIONS - 1)
+        {
+            Thread.yield();
```

Comment:
I suspect we could have an entire conversation just about this line. Does the jmh manual have much to say about this sort of test, where we have to have thread co-ordination because we're testing a concurrent data structure?

Reply:
There's some interesting stuff from Nitsan, [here](http://psy-lob-saw.blogspot.com/p/jmh-related-posts.html) which I'll work my way through. If all else fails I'll have to go and read the full Shipilev lowdown.

Reply:
> jmh manual

I'm not sure such a thing exists? Appears to be some samples in a GitHub project or seeing opinion pieces from various blogs

Reply:
From the quickest of squizzes I think I am prepared to say that there isn't really a consensus, and I get reasonable numbers out of jmh when I run this locally so we're probably good.

Reply:
Out of curiosity, what numbers do you get? As a fresh-faced disruptor-maintainer I wasn't sure what would actually count as "reasonable".

On my i5-6600K I get:

```
Benchmark                                            Mode  Cnt    Score   Error  Units
SimpleBenchmark.publishSimpleEvents                  avgt    5   11.831 ± 0.611  ns/op
SimpleBenchmark.publishSimpleEvents:·pauses          avgt   37  365.617             ms
SimpleBenchmark.publishSimpleEvents:·pauses.avg      avgt         9.882             ms
SimpleBenchmark.publishSimpleEvents:·pauses.count    avgt        37.000              #
SimpleBenchmark.publishSimpleEvents:·pauses.p0.00    avgt         7.692             ms
SimpleBenchmark.publishSimpleEvents:·pauses.p0.50    avgt         9.912             ms
SimpleBenchmark.publishSimpleEvents:·pauses.p0.90    avgt        11.564             ms
SimpleBenchmark.publishSimpleEvents:·pauses.p0.95    avgt        12.306             ms
SimpleBenchmark.publishSimpleEvents:·pauses.p0.99    avgt        13.500             ms
SimpleBenchmark.publishSimpleEvents:·pauses.p0.999   avgt        13.500             ms
SimpleBenchmark.publishSimpleEvents:·pauses.p0.9999  avgt        13.500             ms
SimpleBenchmark.publishSimpleEvents:·pauses.p1.00    avgt        13.500             ms
```

Reply:
Intel(R) Xeon(R) CPU           E5450  @ 3.00GHz

```
Benchmark                                            Mode  Cnt     Score   Error  Units
SimpleBenchmark.publishSimpleEvents                  avgt    5    15.888 ± 2.733  ns/op
SimpleBenchmark.publishSimpleEvents:·pauses          avgt  382  1767.211             ms
SimpleBenchmark.publishSimpleEvents:·pauses.avg      avgt          4.626             ms
SimpleBenchmark.publishSimpleEvents:·pauses.count    avgt        382.000              #
SimpleBenchmark.publishSimpleEvents:·pauses.p0.00    avgt          1.976             ms
SimpleBenchmark.publishSimpleEvents:·pauses.p0.50    avgt          4.184             ms
SimpleBenchmark.publishSimpleEvents:·pauses.p0.90    avgt          7.387             ms
SimpleBenchmark.publishSimpleEvents:·pauses.p0.95    avgt          8.518             ms
SimpleBenchmark.publishSimpleEvents:·pauses.p0.99    avgt         11.092             ms
SimpleBenchmark.publishSimpleEvents:·pauses.p0.999   avgt         14.172             ms
SimpleBenchmark.publishSimpleEvents:·pauses.p0.9999  avgt         14.172             ms
SimpleBenchmark.publishSimpleEvents:·pauses.p1.00    avgt         14.172             ms
```

~100M events/second feels about right for SPSC throughput. Don't ask me where I get that instinct from, though.